### ISSUE TITLE
Web-native loading screen for Flutter Web apps

### ISSUE THREAD

**Issue by orestesgaolin** [flutter/flutter#77936]
## Use case

One common issue with Flutter Web is its loading time (e.g. #46589). Folks tend to blame bundle size, but frankly most of the modern web portals has similar or larger size (famous The V... :P).

Even the Flutter Gallery app experiences similar issue where for couple of seconds black screen is shown.

https://user-images.githubusercontent.com/16854239/110824184-1cce0800-8293-11eb-95f0-1ed16118d16a.mp4

## Proposal

What I would like to propose is to add a web-native (e.g. CSS) loader into the Flutter template for web. This would show other developers that they can improve the UX of their web app by simply showing custom loader. Integrating it with the Flutter app so that it disappears when the bundle is loaded and the app is ready to be rendered.

**Comment by venkatd** [comment]
Agree here. I don't think solution of a loading screen would be a high priority because the blank screen affects end-user perception of the app. They think it's broken.

Meanwhile other web apps (such as Figma) take a similar time to load and have a similar payload size. Because of the loading screen, users don't feel like the app is broken.

**Comment by yjbanov** [comment]
There's two levels of loading screen that I can think of:

1. Have a (potentially animated) indicator that the app is loading.
2. Provide a progress indicator that shows how far along the loading process is.

The first one should already be possible. One can put a full-screen piece of HTML as a placeholder while the app is loading, and register a [post-frame callback](https://api.flutter.dev/flutter/scheduler/SchedulerBinding/addPostFrameCallback.html) that removes the placeholder (ideally using cross-fade). We should provide an example for doing this. It would also be great if our project template included it, so that developers could just take it and hack it into a look according to their app's need.

The second one requires some extra API from Flutter, which we currently do not provide.

**Comment by venkatd** [comment]
@yjbanov even having an example for 1 would go a long way. Progress would be great to have but not critical, while a blank screen causes users to close the tab.

As far as progress, at what layer would that need to get implemented? Is it something that happens at the service worker layer?

**Comment by wilsonowilson** [comment]
This will solve the first issue @yjbanov mentioned:

``` HTML
... 
<style type="text/css">

    .loading {
        display: flex;
        justify-content: center;
        align-items: center;
        margin: 0;
        position: absolute;
        top: 50%;
        left: 50%;
        -ms-transform: translate(-50%, -50%);
        transform: translate(-50%, -50%);
    }

    .loader {
        border: 6px solid #000000;
        border-radius: 50%;
        border-left: 6px solid white;
        width: 40px;
        height: 40px;
        -webkit-animation: spin 1s linear infinite;
        animation: spin 1s linear infinite;
    }

    @-webkit-keyframes spin {
        0% {
            -webkit-transform: rotate(0deg);
        }

        100% {
            -webkit-transform: rotate(360deg);
        }
    }

    @keyframes spin {
        0% {
            transform: rotate(0deg);
        }

        100% {
            transform: rotate(360deg);
        }
    }

</style>
  <title>Planster Admin</title>
  <link rel="manifest" href="manifest.json">
</head>
<body>
  <div class="loading">
    <div class="loader"></div>
  </div>
  <!-- This script installs service_worker.js to provide PWA functionality to
       application. For more information, see:
       https://developers.google.com/web/fundamentals/primers/service-workers -->
  <script>
    if ('serviceWorker' in navigator) {
      window.addEventListener('load', function () {
        navigator.serviceWorker.register('flutter_service_worker.js');
      });
    }
  </script>
```

**Comment by yjbanov** [comment]
For the second scenario I'm imagining Flutter would provide a JavaScript API that pings a callback with the current loading status, something like:

```javascript
window.flutter.addEventListener('load', (status) => {
  status.progress; // a value between 0.0 and 1.0, with 1.0 meaning fully loaded and first frame rendered
  status.action; // action currently being run, such as "canvaskit-fetch", "service-worker-load", "running", etc
});
```

Currently we have no such API. This is partially because our app loading sequence is still evolving. For example, we are planning to begin CanvasKit fetch much earlier in the process.

**Comment by yjbanov** [comment]
... actually `status.action` might need to be `status.actions` as we may be doing multiple things in parallel.

**Comment by gauravmehta13** [comment]
+1

**Comment by Josua2012** [comment]
+1

**Comment by github-bianca** [comment]
Don't forget to remove the element having class "loading"

    if ('serviceWorker' in navigator) {
      window.addEventListener('load', function () {
        var el = document.getElementsByClassName('loading')[0];
        el.remove();
        navigator.serviceWorker.register('flutter_service_worker.js');
      });
    }

**Comment by mrtordoide** [comment]
+1

**Comment by Sunbreak** [comment]
[(#81581) Flutter Web Loading Indicator](https://github.com/flutter/flutter/pull/81581) has been landed since 2.3.0-12.1.pre

**Comment by vistaco** [comment]
if want to remove loading when flutter app loaded just use below code. it's work for me in flutter 2.5.3.

`window.addEventListener("flutter-first-frame",function(){ 
      console.log('flutter app loaded.');
});`

**Comment by iqfareez** [comment]
For recent Flutter version (v3+), we can use the API provided by Flutter in index.html to track the loading process. 

https://docs.flutter.dev/development/platform-integration/web/initialization

But what currently missing is the ability to **track downloading progress** (in percentage perhaps) of canvaskit/main.dart.js file.

Example from Flutter Gallery:

![Flutter web loading splash screen flutter gallery](https://user-images.githubusercontent.com/60868965/178128661-69a21cc0-6e2a-466f-ad61-2e89f9d68862.gif)

**Comment by yjbanov** [comment]
Yes, we launched `flutter.js` in Flutter 3, along with a sample loading animation in the standard template. The API also supports custom interactive landing pages, where, for example, the user can enter their login/password while Flutter is loading in the background. We will continue making incremental improvements from now on, but this issue is largely fixed.

@iqfareez AFAIK on the web there's no way to get percentages for static resources that are loaded over HTTP. This may be possible via WebSockets or other ways of client-server communication. For now, we'll stick with plain HTTP GET fetches that are known to be the fastest and cacheable. Even though that doesn't give us loading progress, the apps will load faster.

**Comment by yjbanov** [comment]
Retroactively assigning to @ditman who actually implemented this.

**Comment by ditman** [comment]
This is how we added a small loading indicator (a pulsing animation, but it can be anything, really) to the gallery, and how it's driven via JS:

* https://github.com/flutter/gallery/commit/8803fd59b8cd566c96ad42febafa39133faf44f6#diff-8f62b6ced28d3396b501d2e89a2e7cb761d16cd7dc977aebece03d4a5da5c24eR77-R95

Note that the code above is slightly outdated, and instead, the promise returning from loadEntrypoint shouldn't be used, instead something like:

```javascript
_flutter.loader.loadEntrypoint({
  serviceWorker: {
    serviceWorkerVersion: serviceWorkerVersion,
  },
  onEntrypointLoaded: _onEntrypointLoaded
});

async function _onEntrypointLoaded(engineInitializer) {
  loading.classList.add('main_done');
  let appRunner = await engineInitializer.initializeEngine();
  loading.classList.add('init_done');
  await appRunner.runApp();
  // Wait a few milliseconds so users can see the "zoooom" animation
  // before getting rid of the "loading" div.
  window.setTimeout(function() {
    loading.remove();
  }, 200); 
}
```

The reason why we switched from the promise-based init to the callback is that first promise was created from javascript, and wouldn't be re-created on hot restart of the app (and the hot restart would break / not work as expected).

Anyway, I hope the idea comes across!

**Comment by ditman** [comment]
I updated the relevant documentation for the 3.7.0 release: 

* https://docs.flutter.dev/development/platform-integration/web/initialization#example-display-a-progress-indicator

I'm closing this as "done", but we can open a new issue if the functionality needs to be expanded.
